### ISSUE TITLE
Reliable Keyspace Notifications

### ISSUE THREAD

**Issue by hpatro** [madolson/placeholderkv#28]
----------------------------
We have a requirement that we need to get a notification on changes to a Redis data structure. Based on my research I found out that I can use Redis key space notifications for doing the same. However, Redis key space notifications send the events to Redis pub/sub channel which is fire and forget i.e once the clients lose the connections all the events till the connection is up again are lost.

Redis streams solve this problem. Also, I want to use consumer group feature of Redis streams. So is there any way that Redis key space notifications can be pushed to Redis streams instead of Redis pub/sub channel?

----------------------------

One of the feature request which I was interested in solving in the past, so filing an issue to have more discussion on this
Ref: https://github.com/redis/redis/issues/5766

**Comment by madolson** [comment]
We had a summary for this didn't we? The conclusion was basically there was no good solution without have a "per-shard hash".

**Comment by hpatro** [comment]
I think we should re-explore and look for solution as we have the opportunity to think through and introduce breaking change with 1.0.

**Comment by CharlesChen888** [comment]
This reminds me of an old request trying to subcribe to a stream (redis/redis#10809), and a reversed request that intended to persist PUBSUB data to streams.

**Comment by xbasel** [comment]
I've marked this for closure and will close the issue soon if there's no further activity.

**Comment by madolson** [comment]
@xbasel we generally don't close feature requests that don't have activity as long as there is still interest.

**Comment by MikaAleksandroff** [comment]
How does one express interest? We would very much like this feature, too.

**Comment by hpatro** [comment]
> How does one express interest? We would very much like this feature, too.

@MikaAleksandroff Could you please explain your use case further and where the current mechanism falls short ?

**Comment by MikaAleksandroff** [comment]
Gladly. 

We have "devices", which send "heartbeats" to our software. Whenever a device stops sending these we need to act on it "immediately" due to certain parameters set by regulations. 

Current way to do that is not very elegant (querying DB for "last heartbeat" for every device and checking if some are overdue. We would like to use Redis TTL; when storing a heartbeat, set TTL for it for X seconds and then just wait for expiry events and process those. 

When I commented on this we were running old Redis (3.x) and ran into issues where those notifications could come half an hour late. I found a video "The evolution of Redis Key Expiration Algorithms", where Salvatore explained the issue and how the routine had been rewritten in v4(?) to fix that. The problem was related to algorithm only scanning some keys and key inactivity played a part and so on..  Now we are finally running Redis 6 and testing again to see if it the modern version really is more reliable. 

It has also become clear that our path forward will be Valkey and now we are also ready to migrate there. I assume Valkey either uses the same algorithms there as Redis 6 or something better and therefore once we finish our current testing we can safely assume the reliability of keyspace notifications will be the same (or even better). Well, now that I have an actual test set up, I can just as well run it against Valkey as well.

We can accept <10 second delay between key expiration and notification, but not more. 

Does that answer your question?

**Comment by MikaAleksandroff** [comment]
@hpatro 

Visual representation of the simulation test running during the weekend. Blue line is between Fri 17 - Mon 08; green line is for Saturday only.

With Actual observed delays (for keys with TTL=60) vary between 1-88 seconds with 99% occurring within 20 seconds of excess delay:
s.	Average	Total
1	0,09 %	0,09 %
2	0,03 %	0,12 %
60	4,30 %	4,42 %
61	8,44 %	12,86 %
62	8,62 %	21,48 %
63	8,77 %	30,25 %
64	8,79 %	39,04 %
65	8,60 %	47,64 %
66	8,24 %	55,88 %
67	7,78 %	63,66 %
68	7,02 %	70,68 %
69	6,01 %	76,69 %
70	5,02 %	81,70 %
71	4,09 %	85,80 %
72	3,32 %	89,12 %
73	2,68 %	91,80 %
74	2,12 %	93,92 %
75	1,63 %	95,56 %
76	1,26 %	96,82 %
77	0,97 %	97,79 %
78	0,74 %	98,53 %
79	0,53 %	99,07 %

Sadly I'm not mathematician let alone statistician.. :-D 

Anyway, to me, this looks like "less than ideal" (with Redis at least), I think for this specific forum I will do the same with the newest Valkey as well.. :-) 

<img width="644" height="363" alt="Image" src="https://github.com/user-attachments/assets/1cf86ba4-ebd3-4ec6-9a07-f1a9f3b78063" />

**Comment by MikaAleksandroff** [comment]
And I just realized that our need is actually a bit different from yours @hpatro as we are mostly worried about notification delay, but not missing any is important, too. :-)  Based on my experiments, the delay issue is fixable by having a thread TOUCHing keys (at a few second intervals), but that perhaps is not very scalable.

**Comment by madolson** [comment]
> We can accept <10 second delay between key expiration and notification, but not more.

Yeah, kind of sounds like deterministic expiration might solve your use case. That will use more memory and compute more CPU, but if you aren't really bottlenecked on those two, we could keep all keys ordered by expiration.

**Comment by hpatro** [comment]
> And I just realized that our need is actually a bit different from yours [@hpatro](https://github.com/hpatro) as we are mostly worried about notification delay, but not missing any is important, too. :-) Based on my experiments, the delay issue is fixable by having a thread TOUCHing keys (at a few second intervals), but that perhaps is not very scalable.

Makes sense. I think you could use the `SCAN` command in the interim (avoid `KEYS` command) within a LUA script, it will increase the CPU utilization by a bit but will ensure timely expiration of keys and further get notification of it. You could delay the entire db scan based on the CPU utilization you can manage.

> Yeah, kind of sounds like deterministic expiration might solve your use case. That will use more memory and compute more CPU, but if you aren't really bottlenecked on those two, we could keep all keys ordered by expiration.

@madolson Do we have thread for this? Or shall we start one?

**Comment by MikaAleksandroff** [comment]
@madolson @hpatro  Thanks guys!

Can't say I know which would be better performance wise, but even with using TOUCH with 32k keys (which I think is about twice the amount required) Redis 6 spent about this much time (from python, batches were 2000 keys each generated within that time):
$ ./test.py 
Connected to master: 10.168.5.120
Fetched 32000 for polling
Elapsed time with batches: 0.6100096702575684 seconds
Fetched 32000 for polling
Elapsed time without: 0.37079644203186035 seconds

And after migration to Valkey 8 with multithreading, blocking perhaps should not be an issue, either. Luckily we don't operate in millions-of-keys space.. :-)

**Comment by madolson** [comment]
> Do we have thread for this? Or shall we start one?

@hpatro Not that I know of. You can start one.
